### PR TITLE
refactor(remove-balances): remove internal balances from accountlib

### PR DESCRIPTION
**Pull request by Alexangelj**
# Description
Since aligning virtual reserves with tracked global balances, removing internal balances entirely significantly reduces the attack surface. It also simplifies the invariant of the contract, which is that the sum of all pool reserves + earned protocol fees = global balance, which is always less than or equal to the actual physical balance of tokens in the contract.

# Changes
- Remove fund, draw, and deposit.
- Remove `balances` mapping from `AccountSystem`  struct.
- Removes protocol fees being credited to internal balance of REGISTRY.
- Adds `claimFee` function for the REGISTRY controller to call.
- Adds `protocolFees` mapping to only track the earned protocol fees.
- Changes the use of `useMax` to use the `getNetBalance` if there is a surplus of tokens in the accounting state, instead of the internal balance of the caller.
- Removes fund, draw, and deposit related events.

### CODE REVIEW

File: contracts/Portfolio.sol [hunk 1]
```diff
@@ -1010,6 +913,28 @@ abstract contract PortfolioVirtual is Objective {
...
+        uint8 decimals = IERC20(token).decimals();
+        if (amount == type(uint256).max) {
+            amountWad = protocolFees[token];
+            amount = amountWad.scaleFromWadDown(decimals);
+        } else {
+            amountWad = amount.scaleToWad(decimals);
+        }
+
+        protocolFees[token] -= amountWad;
+        _decreaseReserves(token, amountWad);
```
Comment (MrToph) [inline contracts/Portfolio.sol:931]: reserves are in WAD now, then it's correct to use `amountWad` here.

File: contracts/Portfolio.sol [hunk 2]
```diff
@@ -247,9 +180,10 @@ abstract contract PortfolioVirtual is Objective {
         PortfolioPair memory pair = pools[poolId].pair;
 
         if (useMax) {
+            // A positive net balance is a surplus of tokens in the accounting state that can be used to mint liquidity.
             deltaLiquidity = pools[poolId].getPoolMaxLiquidity({
-                deltaAsset: getBalance(msg.sender, pair.tokenAsset),
-                deltaQuote: getBalance(msg.sender, pair.tokenQuote)
+                deltaAsset: uint256(getNetBalance(pair.tokenAsset)), // If negative, will be zero.
```
[the remark this answers is not in the document]
Comment (Alexangelj) [reply contracts/Portfolio.sol:185]: Nice catch, fixing

File: contracts/Portfolio.sol [hunk 3]
```diff
@@ -121,28 +122,8 @@ abstract contract PortfolioVirtual is Objective {
...
-        address owner,
-        address token
-    ) public view returns (uint256) {
-        return __account__.balances[owner][token];
-    }
-
     // ===== External Actions ===== //
 
-    /// @inheritdoc IPortfolioActions
-    function deposit() external payable override lock {
```
Comment (MrToph) [inline contracts/Portfolio.sol:135]: this drops support for directly using ETH in the protocol, right? so users need to wrap to WETH first.
But for each WETH pool you're paying out in ETH. Might make sense to completely drop ETH support now?

Comment (Alexangelj) [reply contracts/Portfolio.sol:135]: It's still supported, in the beginning of `multiprocess` the internal `_deposit` is called. This only drops support for carrying balances over between two+ transactions

Comment (MrToph) [reply contracts/Portfolio.sol:135]: right, forgot about the call at the beginning of multiprocess

**Pull request merged**